devito/types/sparse.py

```diff
@@ -338,11 +338,11 @@ def guard(self, expr=None):
                                   condition=condition, indirect=True)
 
         if expr is None:
-            out = self.indexify().xreplace({self._sparse_dim: cd})
+            out = self.indexify()._subs(self._sparse_dim, cd)
         else:
             functions = {f for f in retrieve_function_carriers(expr)
                          if f.is_SparseFunction}
-            out = indexify(expr).xreplace({f._sparse_dim: cd for f in functions})
+            out = indexify(expr).subs({f._sparse_dim: cd for f in functions})
```

Comment:
why subs and not _subs?

and why subs and not xreplace? just performance or because it can be tricky math exprs?

Reply:
>why subs and not _subs?

Multile entries in the dict, `_subs` is for single replacement `_subs(a, b)` 

> and why subs and not replace?

xreplace bit more expensive and end just a bit of extra stuff before calling `subs`.

Note: `indexify` here is ugly but leads to massive issues when removing but for another time.